### ISSUE TITLE
Trying to create static library or native library and testing with C like Rust

### ISSUE THREAD

**Issue by DeafMan1983** [dotnet/runtime#110546]
### Description

Hello,

I want to create classlib as library for static or native like you use Windows dll/lib but I work on Linux Mint with so/a "TestHello" ( project , csproj )

And I test in C programming with header ( "TestHello.h"
and "main.c"

Then it works fine for native library with `gcc -o main main.c ./libTestHello.so -lc -Wl,-Bstat
ic -static-libgcc`

And It looks okay... but libTestHello.a can't compile complete because it throws a lot of lines like undefined references cause Net 8.0 missed to put libcoreclr.a for statical single static-library without worrying required deps...

### Reproduction Steps

Only native library = OK
static library = fail!

### Expected behavior

I have tried....

### Actual behavior

Only native library is working with gcc = successfully!
like
`Hello from C#!` from libTestHello.so, ( `public static unsafe sbyte* ImplHello()` )

With `UnmanagedCallerOnly(EntryPoint="Hello")`

### Known Workarounds

Workaround with native library = OK!

### Configuration

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <SelfContained>true</SelfContained>
    <RuntimeIdentifier>linux-x64</RuntimeIdentifier>
    <PublishAot>true</PublishAot>
    <InvariantGlobalization>true</InvariantGlobalization>
    <AotLinkArgs>--version-script=export.map</AotLinkArgs>
    <StripSymbols>true</StripSymbols>
  </PropertyGroup>
</Project>
```

### Other information

I want to show you how do I create library for native and static library like some Rust- or C/C++ developers create native or static libraries.

```
using System;
using System.Runtime.InteropServices;

public static class MyLibrary
{
    private static readonly sbyte[] HelloMessage = StringToSByteArray("Hello from C#!");

    private static sbyte[] StringToSByteArray(string value)
    {
        if (value == null)
            throw new ArgumentNullException(nameof(value));

        // Allocate buffer with space for the string plus a null terminator
        var buffer = new sbyte[value.Length + 1];

        // Convert each character of the string to sbyte
        for (int i = 0; i < value.Length; i++)
        {
            if (value[i] > sbyte.MaxValue)
                throw new ArgumentException("String contains characters that cannot be represented as sbyte.");

            buffer[i] = (sbyte)value[i];
        }

        // Add null terminator
        buffer[value.Length] = 0;
        return buffer;
    }

    [UnmanagedCallersOnly(EntryPoint = "Hello")]
    public static unsafe sbyte* ImplHello()
    {
        // Return a pointer to the static buffer
        fixed (sbyte* ptr = HelloMessage)
        {
            return ptr;
        }
    }
}
```
And I prompt with **dotnet publish**
`dotnet publish -c Release -p:CppWithCompilerAndLinker=gcc -p:NativeLib=shared`

And locate to native directory and copy TestHello.so to TestHelloForC ( directory for C programming )
And replace "TestHello.so" with "libTestHello.so"!

create header TestHello.h
```
#ifndef TESTHELLO_H
#define TESTHELLO_H

#ifdef __cplusplus
extern "C" {
#endif

const char* Hello();

#ifdef __cplusplus
}
#endif

#endif // TESTHELLO_H
```
And main.c:
```
#include <stdio.h>
#include "TestHello.h" // <- current directory of TestHelloForC

int main()
{
    const char* message = Hello();
    printf("%s\n", message);
    return 0;
}
```
And **gcc ...**:
`gcc -o main main.c ./libTestHello.so -lc -Wl,-Bstatic -static-libgcc`

Result:
```
./main
Hello from C#!
```

But I have problem with static-library :(

cause error throws a lot of `undefined references?`

Like I show you a lot of errors during I already published as static version of library and try to compile with gcc and gcc throws a lot of errors...
![Image](https://github.com/user-attachments/assets/51c86c07-2f09-4bff-82ce-c52d94de2b78)
Too large lines :/

How do I know if I want to make sure static linking from Dotnet publish?

If Dotnet 8.0 missed to put libcoreclr.a or doesn't not exist in runtime-ilcompile package?

Thanks!

**Comment by jkotas** [comment]
We recommend to use shared libraries (.so on Linux). They are much easier to build and use.

Building static libraries is very advanced scenario. We have it briefly documented at https://github.com/dotnet/samples/blob/main/core/nativeaot/NativeLibrary/README.md#building-static-libraries . You have to find and copy the full linker used to link shared libraries over to your project.

**Comment by AaronRobinsonMSFT** [comment]
Note that the building of and consumption of static libraries through native AOT is not officially supported - see https://learn.microsoft.com/dotnet/core/deploying/native-aot/libraries.

As mentioned aspects of it are documented, but it is a very advanced feature and not something we generally provide support for.

**Comment by DeafMan1983** [comment]
Okay thanks for explanation. You mean Net 8 and 9 are still beta version cause he said...
> Note that the building of and consumption of static libraries through native AOT is not officially supported.

That is sadly. Since I used starting version Net 7 or 8 but I don't know since I already embedded X11 into Dotnet executable = OK.
static library = OK but before I have made.. But I forget since when I made...

Thanks for everything!
I feel shy because Rust developers are happy and try to make static libraries against Net developers. It is really painfully.

Come on, M$ Net developers! - I think I will close it.
